server/src/index.ts

```diff
@@ -3,17 +3,81 @@
 // SPDX-License-Identifier: MIT
 
 import express, { Application, Request, Response } from 'express';
+import { validateLogin, performSignup, PlainPassword, UserRole } from './auth';
 import { TeacherCourses, getTeacherCourses } from './services/courses';
 import cors from 'cors';
 
 const app: Application = express();
 const parsedPort = Number(process.env.AALTO_GRADES_BACKEND_PORT);
 const port: number = isNaN(parsedPort) ? 3000 : parsedPort;
 
+interface LoginRequest {
+  username: string,
+  password: PlainPassword,
+}
+
+interface SignupRequest {
+  username: string,
+  password: PlainPassword,
+  email: string,
+  role: UserRole,
+}
+
+function validateUserRole(role: any): role is UserRole {
+  return typeof role === 'string' && (
+    role === 'Teacher' ||
+    role === 'Student' ||
+    role === 'Admin'
+  );
+}
+
+function validateLoginFormat(body: any): body is LoginRequest {
+  return body &&
+    body.username &&
+    body.password &&
+    typeof body.username === 'string' &&
+    typeof body.password === 'string';
+}
+
+function validateSignupFormat(body: any): body is SignupRequest {
+  return body &&
+    body.username &&
+    body.password &&
+    body.email &&
+    validateUserRole(body.role) &&
+    typeof body.username === 'string' &&
+    typeof body.password === 'string' &&
+    typeof body.email === 'string';
```

Comment:
Why are these type checks here? These seem like things that would already be ensured by `body is SignupRequest`.

Reply:
`body is SignupRequest` is a way of indicating that we have checked that `body` is of the correct type/interface. It's basically a hint to the type checking engine, and it doesn't have a runtime effect.